### PR TITLE
Cross build for SBT 1.0.0-RC3

### PR DESCRIPTION
**Pull request by liff**
Uses scalariform for 2.12 from `com.github.machaval` organization (see #55 and #36) so perhaps this is not the final form.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -14,18 +14,24 @@ developers := List(
...
+) ++ (if (scalaVersion.value startsWith "2.10.") List("-target:jvm-1.6") else List.empty)
 
 resolvers ++= Seq(sonatypeSnapshots, sonatypeReleases)
 
-libraryDependencies += "org.scalariform" %% "scalariform" % "0.1.8"
+libraryDependencies += {
+  if (scalaVersion.value startsWith "2.10.")
+    "org.scalariform"     %% "scalariform" % "0.1.8"
+  else
+    "com.github.machaval" %% "scalariform" % "0.2.0"
```
Comment (godenji) [inline build.sbt]: > "org.scalariform" %% "scalariform" % "0.2.0"

Has been released, includes changes from the `com.github.machaval` fork, cross published for 2.10/11/12.

Please amend PR accordingly. Thanks

File: build.sbt [hunk 2]
```diff
@@ -14,18 +14,24 @@ developers := List(
...
   "-encoding", "UTF-8"
-)
+) ++ (if (scalaVersion.value startsWith "2.10.") List("-target:jvm-1.6") else List.empty)
 
 resolvers ++= Seq(sonatypeSnapshots, sonatypeReleases)
 
-libraryDependencies += "org.scalariform" %% "scalariform" % "0.1.8"
+libraryDependencies += {
+  if (scalaVersion.value startsWith "2.10.")
+    "org.scalariform" %% "scalariform" % "0.1.8"
```
Comment (godenji) [inline build.sbt]: thanks, the reason for mentioning cross publishing is that library dependencies can now simply be:

`libraryDependencies += "org.scalariform" %% "scalariform" % "0.2.0"`

Comment (liff) [inline build.sbt]: Yeah, seems I failed to properly parse your last commend, sorry. Update on its way.

I was trying to look for [0.2.0 for 2.10](http://search.maven.org/#search%7Cga%7C1%7Ca%3Ascalariform_2.10) on Maven Central but I suppose its indexing hasn't caught up yet or something.

Comment (godenji) [inline build.sbt]: right, it's not yet up on maven central, but the dependency resolves (0.2.0 was published an hour or so ago)

**Comment by godenji** [comment]
@liff thanks for the update. 

Scala 2.12.3 will soon be released (likely within days), I'd like to hold off on publishing a new `sbt-scalariform` release until that time (we're also waiting on a PR that brings embedded scalariform in Scala IDE up-to-date with latest version of scalariform).

More later.

**Comment by godenji** [comment]
@liff thanks for the contribution, will be part of the 1.8.0 release (probably in a couple of weeks when sbt 1.0 lands) .

**Comment by liff** [comment]
From my understanding of the [announcement](http://developer.lightbend.com/blog/2017-07-28-sbt-1-0-0-RC3-and-sbt-0-13-16/index.html) you don't necessarily have to wait for 1.0 final to release for it.

**Comment by godenji** [comment]
Sure, but latest changes need to be vetted before releasing 1.8.0 to make sure everything is working properly, which is not the case right now, seems to be broken.

Did you try out 1.0.0-RC3 + scala 2.12 against PR changes?  Everything's fine for 0.13.16 + 2.10.6, but when doing a `^publishLocal` from `sbt-scalariform` and creating a sample sbt project with:

`build.properties` > 1.0.0-RC3
`project/plugins.sbt` >  `addSbtPlugin("org.scalariform" % "sbt-scalariform" % "1.8.0-SNAPSHOT")`
`src/main/scala` > Sample.scala

formatting fails with:

> [error] jawn.IncompleteParseException: exhausted input
[error] 	at jawn.Parser.parse(Parser.scala:362)
[error] 	at jawn.SyncParser.parse(SyncParser.scala:24)
[error] 	at jawn.SupportParser.parseUnsafe(SupportParser.scala:12)
[error] 	at jawn.SupportParser.parseUnsafe$(SupportParser.scala:11)
[error] 	at sjsonnew.support.scalajson.unsafe.Parser$.parseUnsafe(Parser.scala:8)
[error] 	at sbt.util.CacheStore$.$anonfun$jvalueIsoString$2(CacheStore.scala:17)
[error] 	at sjsonnew.IsoString$$anon$1.from(IsoString.scala:27)
[error] 	at sbt.util.PlainInput.read(Input.scala:29)
[error] 	at sbt.util.FileBasedStore.$anonfun$read$1(CacheStore.scala:61)
[error] 	at sbt.io.Using.apply(Using.scala:23)
[error] 	at sbt.util.FileBasedStore.read(CacheStore.scala:61)
[error] 	at sbt.util.Changed.uptodate(Tracked.scala:225)
[error] 	at sbt.util.Changed.$anonfun$apply$1(Tracked.scala:214)
[error] 	at com.typesafe.sbt.Scalariform$.apply(Scalariform.scala:62)
...

**Comment by liff** [comment]
I didn't try actually. Assumed it'd just work. I'll check it out in the evening.

**Comment by godenji** [comment]
> Assumed it'd just work

Famous last words ;-)

Think your code is fine, looks like real cause of the exception may be upstream in `sjsonnew`. There are at least two problems that I can see. 

First, in sbt 1.0 the empty file case seems to not be properly encoded by sjsonnew for `jawn` to consume. See [sbt 1.0 Changed[0]](https://github.com/sbt/util/blob/1.x/util-tracking/src/main/scala/sbt/util/Tracked.scala#L210) vs. [sbt 0.13.16 Changed[0]](https://github.com/sbt/sbt/blob/v0.13.16/cache/tracking/src/main/scala/sbt/Tracked.scala#L148). In the latter the `uptodate` method catches IO exception whereas the former optimistically assumes that the (potentially empty) file can be read in as the target type.

Second, to workaround the first problem, even when stuffing dummy data (like "alignParameters=false") into `target/sbt-scalariform/streams/compile/scalariformFormat/\$global/streams/scalariform-preferences`, the data is for some reason not cached -- `compile` and `scalariformFormat` format sources over and over despite the preferences not having been changed.

Please try to sort this out, will likely involve modifying your code to workaround upstream issues. If you issue a new PR, please do so against my fork in the [1.8.0 branch](https://github.com/godenji/sbt-scalariform/tree/1.8.0), have some new functionality built on top of current master that will go into 1.8.0 release. That or fixup your fork and I'll clone it to test out changes prior to your issuing a PR to master.

p.s. you can run `^scripted` from my fork on 1.8.0; if the tests pass then we're good.

Thanks

**Comment by liff** [comment]
Opened a PR about the caching issue. I think the exception handling thing might be just a regression in SBT 1.0. I'll ask around in gitter tomorrow or open an issue.

**Comment by godenji** [comment]
We're getting there, your PR to my fork solves the caching issue. Now we need to sort out the empty cache file scenario (which is always the case on fresh project, or after an sbt `clean`).

SBT 1.0 final is scheduled to be released in a couple of weeks. Hopefully they can be more [defensive here](https://github.com/sbt/util/blob/1.x/util-tracking/src/main/scala/sbt/util/Tracked.scala#L225). That, or we try to workaround the issue by checking the empty file case on our end.

**Comment by liff** [comment]
It looks like this is going to be more complicated than even fixing the SBT regression. `Changed` is going to be deprecated but apparently without fixing the exception behavior in 1.0.0.

The suggested replacement is to use `Tracked.inputChanged` but its innards have also changed beween 0.13 and 1.0. In 0.13 the `Equiv` provided for the cached type is [used](https://github.com/sbt/sbt/blob/6ce09cfed6f3353326981a4e6e07ffba1f15252f/cache/tracking/src/main/scala/sbt/Tracked.scala#L110) to find out if the object changed. In 1.0 the logic has been changed to [use hash code](https://github.com/sbt/util/blob/7b02a40b5ec0840fd5d7ca40d920e4d6ffc6419f/util-tracking/src/main/scala/sbt/util/Tracked.scala#L175). Due to the somewhat curious way [`Equiv` is defined](https://github.com/sbt/sbt-scalariform/blob/61f6aabf18680cf7061d01f29b7351b0bee74709/src/main/scala/com/typesafe/sbt/Scalariform.scala#L97) for `IFormattingPreferences` (i.e. empty Map and defaults are equivalent), using `inputChanged` doesn't really work in 1.0 either.

I suppose one option could be to always write out the preferences in full so that the `Equiv` trick is not needed. What do you think?

**Comment by godenji** [comment]
Agreed, 0.13 to 1.0 is full of pitfalls (dealing with similar problems trying to port `sbt-eclipse` to 1.0). 

I have tried with `Tracked.inputChanged` (see below), always returns `true` regardless of whether or not formatting preferences have changed. I am not sure what is happening here, ideally we could use `inputChanged` instead of the following workaround:

Use our own internal definition of `Changed` with try/catch clause added (works, have in place on local branch), and then remove it when 0.13 support is dropped. At that point everything can be moved back under `src/main/scala` and a 1.0-specific implementation can be used.

`Tracked.inputChanged` attempt
```
private[sbt] class Changed[O: Equiv: JsonFormat](store: File) {

  def apply(ifChanged: O => Boolean, ifUnchanged: O => Boolean): O => Boolean = value => {
      val f = Tracked.inputChanged[O, Boolean](store) { case (changed, _) =>
        changed
      }   
      f(value)
    }   
  }

  def changed[O: Equiv: JsonFormat](store: File) = new Changed[O](CacheStore(store))
```

**Comment by liff** [comment]
Yeah, `Tracked.inputChanged` doesn't work because (from what I can tell, check the links in my post above) it doesn't really use the `Equiv` instance in 1.0.

Using a custom `Changed` implementation is something I tried as well earlier but deemed a bit too hacky. However, having gone through all this, it's starting to look like the easiest option :)

**Comment by godenji** [comment]
Life is too short, let's just get this PR fixed up and deal with perfection later (i.e. when sbt 0.13 can be dropped). Unless you have time that is, feel free to try to get `inputChanged` working alongside 0.13 if you like ;-)

I pushed patched `Changed` implementation to my fork on 1.8.0 branch. When you get the chance can you pull my latest changes and confirm that `^scripted` tests pass for both 0.13.16 and 1.0.0-RC3?

**Comment by godenji** [comment]
@liff Tests pass, think we're good to go for 1.8.0

**Comment by liff** [comment]
Sorry, I was busy with other stuff the past couple of days.

**Pull request merged**